### PR TITLE
Add page finishing function when over scrolled to SessionDetailActivity

### PR DESCRIPTION
**Pull request by chibatching**
## Issue
- https://github.com/DroidKaigi/conference-app-2017/issues/228

## Overview (Required)
- Finish SessionDetailActivity when page overscrolled

## Links
- none

## Screenshot
Before | After
:--: | :--:
 |

### CODE REVIEW

**Comment by konifar** [comment]
Cool 🆒

**Comment by konifar** [comment]
Interesting implementation 👀 
I thought it's better to make custom activity, but your custom CoordinatorLayout also looks good and reasonable. Let me taste this on my device after CI success 👍

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import android.support.v4.widget.NestedScrollView;
+import android.util.AttributeSet;
+import android.view.View;
+import android.view.animation.AccelerateDecelerateInterpolator;
+
+import io.github.droidkaigi.confsched2017.R;
+
+
+@BindingMethods({
+        @BindingMethod(type = OverScrollLayout.class, attribute = "onOverScroll", method = "setOverScrollListener")
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: :cool:

**Comment by konifar** [comment]
Sorry to be late review 🙇 
I checked on my device. It's really awesome! Thank you so much 👍 
LGTM! 💯

**Comment by chibatching** [comment]
Thank you for your review!! 😄

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Override
+    public boolean onNestedPreFling(View target, float velocityX, float velocityY) {
+        return getY() != originalLocation.y || super.onNestedPreFling(target, velocityX, velocityY);
+    }
+
+    @Override
+    public void onStopNestedScroll(View target) {
+        super.onStopNestedScroll(target);
+        if (Math.abs(getY()) > overScrollThreshold && overScrollListener != null) {
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: `overScrollListener != null` should be checked first.
Could you please swap?

Comment (chibatching) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: Oh, I see. I'd like to make another PR to fix it.

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    private static final float OVER_SCROLL_THRESHOLD = 0.20f;
+
+    private static final int RESTORE_ANIM_DURATION = 100;
+
+    private static final float MINIMUM_OVER_SCROLL_SCALE = 0.99f;
+
+    private int overScrollThreshold;
+
+    private PointF originalLocation = new PointF();
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: `Rect` is better, I think. Then, Rect#getHeight can be used instead of originalHeight.

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+import io.github.droidkaigi.confsched2017.R;
+
+
+@BindingMethods({
+        @BindingMethod(type = OverScrollLayout.class, attribute = "onOverScroll", method = "setOverScrollListener")
+})
+public class OverScrollLayout extends CoordinatorLayout {
+
+    private static final float OVER_SCROLL_THRESHOLD = 0.20f;
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: This variable is a scale or ratio to calculate threshold?
Can you add any suitable suffix?

File: app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java [hunk 5]
```diff
@@ -157,6 +157,10 @@ public void onClickFab(@SuppressWarnings("unused") View view) {
         }
     }
 
+    public void onOverScroll() {
+        callback.onOverScroll();
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java]: should check null or not.

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                    .scaleY(1);
+        }
+    }
+
+    private void translationY(float dy) {
+        setY(getY() + dy * 0.5f);
+    }
+
+    private boolean isAppBarExpanded(@NonNull AppBarLayout appBarLayout) {
+        return appBarLayout.getHeight() == appBarLayout.getBottom();
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: Does this really work?
View#getHeight() is calculated by View#getBottom() - View#getTop().
Are isAppBarExpanded and View#getTop() == zero used equally?

**Comment by jmatsu** [comment]
Oops, I didn't notice this PR was merged 😇

**Comment by chibatching** [comment]
I'd like to make another PR to fix your commet! 👍

**Comment by konifar** [comment]
@jmatsu Thanks for great review 😄 
@chibatching Appreciate you. Looking forward to next PR 👍

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java (hunk 6 above)
Comment (chibatching) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/OverScrollLayout.java]: > View#getHeight() is calculated by View#getBottom() - View#getTop()

I didn't know this, it may be able to replaced with View#getTop() == zero.
Thanks!
